File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int vis[n][m];
        int cntFresh=0;
        queue<pair<pair<int,int>,int>> q;
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                if(grid[i][j]==2) {
                    q.push({{i,j},0});
                    vis[i][j]=2;
                }else {
                    vis[i][j]=0;
                }
                if(grid[i][j]==1) cntFresh++;
            }
        }
        
        int delRow[]={-1,0,1,0};
        int delCol[] = {0,1,0,-1};
        int cnt=0;      
        int tm =0;
        while(!q.empty()) {
            int row = q.front().first.first;
            int col = q.front().first.second;
            int t = q.front().second;
            tm = max(tm,t);
            q.pop();
            for(int i=0;i<4;i++) {
                int nrow=row+delRow[i];
                int ncol=col+delCol[i];
                if(nrow>=0 && nrow<n && ncol>=0 && ncol<m && vis[nrow][ncol]==0 && grid[nrow][ncol]==1) {
                    vis[nrow][ncol]=2;
                    q.push({{nrow,ncol},t+1});
                    cnt++;
                }
            }
        }
        if(cnt!=cntFresh) return -1;
        return tm;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;

        int vis[][] = new int[n][m];
        Queue<int[]> q = new LinkedList<>();
        int cntFresh=0;
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                if(grid[i][j]==2) {
                    q.offer(new int[]{i,j,0});
                    vis[i][j]=1;
                }else if(grid[i][j]==1) {
                    cntFresh++;
                }else {
                    vis[i][j]=0;
                }
            }
        }

        int delRow[] = new int[]{-1,0,+1,0};
        int delCol[] = new int[]{0,+1,0,-1};
        int tm = 0;

        while(!q.isEmpty()) {
            int row = q.peek()[0];
            int col = q.peek()[1];
            int t = q.peek()[2];
            q.poll();
            tm = Math.max(tm, t);
            for(int i=0;i<4;i++) {
                int nrow = row+delRow[i];
                int ncol = col+delCol[i];
                if(nrow>=0 && nrow<n && ncol>=0 && ncol<m
                 && grid[nrow][ncol]==1&&vis[nrow][ncol]==0) {
                    vis[nrow][ncol]=1;
                    cntFresh--;
                    grid[nrow][ncol]=2;
                    q.offer(new int[]{nrow,ncol,tm+1});
                 }
            }
        }

        if(cntFresh!=0) return -1;
        return tm;
    }
}